File: scripts/fetch-hcstats.mjs
// Node 20+ — Fetch Hockey Report stats and write /public/hcstats.json

import fs from "node:fs/promises";
import path from "node:path";

const TOKEN = process.env.HC_REPORT_TOKEN;
const TEAM_ID_RAW = process.env.HC_REPORT_TEAM_ID;
const SEASON_ID_RAW = process.env.HC_REPORT_SEASON_ID;
const PHASE_ID_RAW = process.env.HC_REPORT_PHASE_ID;

if (!TOKEN || !TEAM_ID_RAW) {
  console.error("Missing required envs: HC_REPORT_TOKEN, HC_REPORT_TEAM_ID.");
  process.exit(1);
}

const TEAM_ID = Number(TEAM_ID_RAW);
if (!Number.isFinite(TEAM_ID)) {
  console.error("HC_REPORT_TEAM_ID must be a valid number.");
  process.exit(1);
}

const SEASON_ID = SEASON_ID_RAW ? Number(SEASON_ID_RAW) : undefined;
if (SEASON_ID_RAW && !Number.isFinite(SEASON_ID)) {
  console.error("HC_REPORT_SEASON_ID must be a valid number.");
  process.exit(1);
}

const PHASE_ID = PHASE_ID_RAW ? Number(PHASE_ID_RAW) : undefined;
if (PHASE_ID_RAW && !Number.isFinite(PHASE_ID)) {
  console.error("HC_REPORT_PHASE_ID must be a valid number.");
  process.exit(1);
}

const API = "https://hockey-report.eu/api";
const payload = { token: TOKEN, team_id: TEAM_ID };

if (SEASON_ID !== undefined) payload.season_id = SEASON_ID;
if (PHASE_ID !== undefined) payload.phase_id = PHASE_ID;

async function postJson(endpoint) {
  const res = await fetch(`${API}/${endpoint}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload)
  });
  const text = await res.text();
  if (!res.ok) {
    throw new Error(`${endpoint}: ${res.status} ${res.statusText} - ${text}`);
  }
  const data = JSON.parse(text);
  if (data && data.error) {
    throw new Error(`${endpoint}: ${data.error}`);
  }
  return data;
}

(async () => {
  const [table, games, players] = await Promise.all([
    postJson("api_json_table.php"),
    postJson("api_json_games.php"),
    postJson("api_json_top_players.php")
  ]);

  const output = {
    updated_at: new Date().toISOString(),
    team_id: TEAM_ID,
    context: table?.context || games?.context || players?.context || {},
    standings: table?.standings || [],
    games: games?.games || [],
    players: players?.players || []
  };

  const outDir = path.join(process.cwd(), "public");
  await fs.mkdir(outDir, { recursive: true });
  await fs.writeFile(path.join(outDir, "hcstats.json"), JSON.stringify(output, null, 2) + "\n", "utf8");

  console.log("Wrote public/hcstats.json");
})().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: python/fetch_hcstats.py
# Python 3.12+ - Fetch Hockey Report stats and write /public/hcstats.json
import argparse
import asyncio
import json
import os
import time
from datetime import datetime
from pathlib import Path

import httpx
from dotenv import load_dotenv

load_dotenv()

API_BASE_URL = os.environ.get("HC_REPORT_API_URL")
TOKEN = os.environ.get("HC_REPORT_TOKEN")
SCRIPT_DIR = Path(__file__).resolve().parent
REPO_ROOT = SCRIPT_DIR.parent
DEFAULT_CONFIG_PATH = SCRIPT_DIR / "config_season.json"
OUTPUT_FILE_PATH = REPO_ROOT / "public" / "hcstats.json"
REQUIRED_QUERY_FIELDS = ("team_id", "league_id", "division_id", "season_id", "phase_id")
CONTEXT_FIELDS = ("league_id", "division_id", "season_id", "phase_id")


def parse_args():
    parser = argparse.ArgumentParser(
        description="Fetch Hockey Report stats and write public/hcstats.json."
    )
    parser.add_argument(
        "--config",
        default=str(DEFAULT_CONFIG_PATH),
        help=(
            "Path to query config JSON with fields: team_id, league_id, "
            "division_id, season_id, phase_id."
        ),
    )
    parser.add_argument(
        "--strict-context",
        action="store_true",
        help=(
            "Fail if API response context (season/phase/league/division) "
            "does not match config values."
        ),
    )
    return parser.parse_args()


def load_query_config(config_path):
    try:
        with open(config_path, "r", encoding="utf8") as config_file:
            raw_config = json.load(config_file)
    except FileNotFoundError as err:
        raise ValueError(f"Config file not found: {config_path}") from err
    except json.JSONDecodeError as err:
        raise ValueError(f"Invalid JSON in config file {config_path}: {err}") from err

    config = {}
    for field in REQUIRED_QUERY_FIELDS:
        if field not in raw_config:
            raise ValueError(f"Missing required config field: {field}")
        try:
            config[field] = int(raw_config[field])
        except (ValueError, TypeError) as err:
            raise ValueError(f"Config field {field} must be a valid number.") from err
    return config


def parse_int(value):
    try:
        return int(value)
    except (ValueError, TypeError):
        return None


def extract_context(payload):
    if not isinstance(payload, dict):
        return {}
    raw_context = payload.get("context")
    if not isinstance(raw_context, dict):
        return {}
    context = {}
    for field in CONTEXT_FIELDS:
        parsed = parse_int(raw_context.get(field))
        if parsed is not None:
            context[field] = parsed
    return context


def find_context_mismatches(endpoint_contexts, query_config):
    mismatches = {}
    for endpoint, context in endpoint_contexts.items():
        if not context:
            continue
        endpoint_mismatches = {}
        for field in CONTEXT_FIELDS:
            actual = context.get(field)
            expected = query_config.get(field)
            if actual is None or expected is None:
                continue
            if actual != expected:
                endpoint_mismatches[field] = {"expected": expected, "actual": actual}
        if endpoint_mismatches:
            mismatches[endpoint] = endpoint_mismatches
    return mismatches


async def post_json(client, endpoint, api_base_url, payload):
    res = await client.post(
        f"{api_base_url}/{endpoint}",
        params={"_": int(time.time() * 1000)},
        json=payload,
        headers={
            "Content-Type": "application/json",
            "Cache-Control": "no-cache",
            "Pragma": "no-cache",
        },
    )
    text = res.text
    if not res.is_success:
        raise Exception(f"{endpoint}: {res.status_code} {res.reason_phrase} - {text}")
    data = res.json()
    if data and data.get("error"):
        raise Exception(f"{endpoint}: {data['error']}")
    return data


async def main():
    args = parse_args()

    if not API_BASE_URL or not TOKEN:
        print("Missing required envs: HC_REPORT_API_URL, HC_REPORT_TOKEN.")
        return 1

    try:
        query_config = load_query_config(args.config)
    except ValueError as err:
        print(err)
        return 1

    payload = {"token": TOKEN, **query_config}

    async with httpx.AsyncClient() as client:
        table, games, players = await asyncio.gather(
            post_json(client, "api_json_table.php", API_BASE_URL, payload),
            post_json(client, "api_json_games.php", API_BASE_URL, payload),
            post_json(client, "api_json_top_players.php", API_BASE_URL, payload),
        )

    endpoint_contexts = {
        "table": extract_context(table),
        "games": extract_context(games),
        "players": extract_context(players),
    }
    context_mismatches = find_context_mismatches(endpoint_contexts, query_config)
    if context_mismatches:
        print(f"Warning: API context differs from config: {json.dumps(context_mismatches)}")
        if args.strict_context:
            return 1

    output = {
        "updated_at": datetime.now().isoformat(),
        "team_id": query_config["team_id"],
        "context": {field: query_config[field] for field in CONTEXT_FIELDS},
        "response_contexts": endpoint_contexts,
        "standings": table.get("standings") or [],
        "games": games.get("games") or [],
        "players": players.get("players") or [],
    }

    OUTPUT_FILE_PATH.parent.mkdir(parents=True, exist_ok=True)
    with OUTPUT_FILE_PATH.open("w", encoding="utf8") as f:
        json.dump(output, f, indent=2)
        f.write("\n")

    print(f"Wrote {OUTPUT_FILE_PATH}")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(asyncio.run(main()))
    except Exception as e:
        print(e)
        exit(1)
